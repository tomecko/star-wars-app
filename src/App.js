import React from 'react';

import './App.css';

function App() {
  return (
    <div>
      Star Wars App
    </div>
  );
}

export default App;
