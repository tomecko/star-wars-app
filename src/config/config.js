export const BATTLE_TIMEOUT = 3000;
export const ITEM_COUNT = 2;

export const RESOURCE_INFOS = [
  {
    name: 'people',
    getScore: ({ mass }) => Number(mass) || 0,
    getScoreText: ({ mass }) => mass,
    scoreDescription: 'mass',
  },
  {
    name: 'starships',
    getScore: ({ crew }) => Number(crew) || 0,
    getScoreText: ({ crew }) => crew,
    scoreDescription: 'crew',
  },
];
