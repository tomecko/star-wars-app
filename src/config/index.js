export * from './config';
export * from './machine';
