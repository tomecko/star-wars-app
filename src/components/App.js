import React, { useEffect } from 'react';
import { Container, Row, Col } from 'reactstrap';
import { useMachine } from '@xstate/react';

import { machine } from '../config';

import { Battle } from './Battle';
import { Failure } from './Failure';
import { Footer } from './Footer';
import { Header } from './Header';
import { Loading } from './Loading';
import { Menu } from './Menu';
import { Scores } from './Scores';

export function App(props) {
  const [current, send] = useMachine(props.machine || machine);
  useEffect(() => {
    send('INIT');
  }, [send])

  return (
    <Container>
      <Row>
        <Col xs="12" md="8" lg="6">
          <Header />
          <Scores scores={current.context.scores} />

          {current.matches('loading') ? <Loading /> : ''}

          {current.matches('failure') ? (
            <Failure error={current.context.error} send={send} />
           ) : ''}

          {current.matches('menu') ? (
            <Menu resourceInfo={current.context.resourceInfo} send={send} />
           ) : null}

          {current.matches('battle') && current.context.battle
            ? <Battle
                battle={current.context.battle}
                resourceInfo={current.context.resourceInfo}
              />
            : null}

          <Footer />
        </Col>
      </Row>
    </Container>
  );
}

