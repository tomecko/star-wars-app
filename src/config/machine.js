import { Machine, assign } from 'xstate';

import { getNewScores, getZeroes, getNewBattle } from '../helpers';
import { starWarsDataService } from '../services';

import { BATTLE_TIMEOUT, ITEM_COUNT, RESOURCE_INFOS } from './config';

export const machine = Machine({
  context: {
    battle: null,
    resourceInfo: null,
    scores: getZeroes(ITEM_COUNT),
  },
  initial: 'blank',
  states: {
    blank: {
      on: {
        INIT: 'loading',
      },
    },
    loading: {
      invoke: {
        src: 'loadingInvoke',
        onDone: 'menu',
      },
    },
    menu: {
      on: {
        PLAY: {
          target: 'battle',
          actions: assign({ resourceInfo: (_, event) => event.resourceInfo }),
        },
      }
    },
    battle: {
      invoke: {
        src: 'battleInvoke',
        onDone: {
          actions: assign(({
            battle: getNewBattle,
            scores: getNewScores,
          })),
        },
      },
      after: {
        [BATTLE_TIMEOUT]: 'menu',
      },
      exit: assign(({ battle: null })),
    },
  },
}, {
  services: {
    loadingInvoke: () => Promise.all(RESOURCE_INFOS.map(({ name }) =>
      starWarsDataService.getAll(name))),
    battleInvoke: (context) =>
      starWarsDataService.getRandom(context.resourceInfo.name, ITEM_COUNT),
  },
});
