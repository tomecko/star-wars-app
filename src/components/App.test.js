import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { App } from './App';
import { machine } from '../config';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches snapshot', () => {
  const component = renderer.create(
    <App />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('allows playing a game', () => {
  const div = document.createElement('div');
  const mockMachine = machine.withConfig({
    services: {
      loadingInvoke: () => Promise.resolve(),
      battleInvoke: () => Promise.resolve([
        { name: 'name1', a: 1 },
        { name: 'name2', a: 10 },
      ]),
    },
  });
  ReactDOM.render(<App machine={mockMachine} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
