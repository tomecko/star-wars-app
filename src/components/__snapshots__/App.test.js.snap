// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-12 col-md-8 col-lg-6"
    >
      <div
        className="jumbotron"
      >
        <h1>
          Star Wars App
        </h1>
        <p>
          Are you ready for the action?
        </p>
      </div>
      <div
        className="card card-body"
      >
        <div>
          Player 
          1
          <div
            className="progress"
          >
            <div
              aria-valuemax={10}
              aria-valuemin="0"
              aria-valuenow={0.5}
              className="progress-bar"
              role="progressbar"
              style={
                Object {
                  "width": "5%",
                }
              }
            >
              0
            </div>
          </div>
        </div>
        <div>
          Player 
          2
          <div
            className="progress"
          >
            <div
              aria-valuemax={10}
              aria-valuemin="0"
              aria-valuenow={0.5}
              className="progress-bar"
              role="progressbar"
              style={
                Object {
                  "width": "5%",
                }
              }
            >
              0
            </div>
          </div>
        </div>
      </div>
      
      
      <footer>
        <div
          className="card card-body"
        >
          <ul>
            <li>
              For demonstration purposes fetching data fails with 50% probability on the first attempt.
            </li>
            <li>
              Currently the app loads all data on startup. It does not scale and may take some time especially if SW API is slow :( Alternatively it would be possible to get random pair of ids on the client side and then fetch two items by their id. That would be correct only if made strong assumptions about all ids (i.e. ids are always subsequent integers). On the other hand the current solution (downloading all data) results in smooth and instantaneous interactions during the game.
            </li>
            <li>
              As a challenge I decided to try out a new approach toward handling app logic and configured state machine (XState). Naturally it would also be possible to set up Redux store or just use component's state as the app is relatively simple.
            </li>
          </ul>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
